File: 004_CanPlaceFlowers.py
"""
Topic: Greedy / Array
Can Place Flowers: check if we can plant n flowers without adjacent planting
Time Complexity: O(n)
Space Complexity: O(1)
"""

def can_place_flowers(flowerbed, n):
    i = 0
    while i < len(flowerbed) and n > 0:
        if flowerbed[i] == 0:
            prev_empty = (i == 0) or (flowerbed[i-1] == 0)
            next_empty = (i == len(flowerbed)-1) or (flowerbed[i+1] == 0)
            if prev_empty and next_empty:
                flowerbed[i] = 1
                n -= 1
                i += 1
        i += 1
    return n == 0


if __name__ == '__main__':
    print(can_place_flowers([1,0,0,0,1], 1))  # True
    print(can_place_flowers([1,0,0,0,1], 2))  # False

File: 004_CanPlaceFlowers.cpp
// Topic: Greedy / Array
// Can Place Flowers: Determine if n new flowers can be planted in a flowerbed without adjacent
// planting. Time Complexity: O(m) where m = flowerbed.size() Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

bool canPlaceFlowers(vector<int>& flowerbed, int n)
{
    int m = flowerbed.size();
    for (int i = 0; i < m && n > 0; ++i)
    {
        if (flowerbed[i] == 0)
        {
            bool left = (i == 0) || (flowerbed[i - 1] == 0);
            bool right = (i == m - 1) || (flowerbed[i + 1] == 0);
            if (left && right)
            {
                flowerbed[i] = 1;
                --n;
            }
        }
    }
    return n <= 0;
}

int main()
{
    vector<int> f1 = {1, 0, 0, 0, 1};
    cout << (canPlaceFlowers(f1, 1) ? "true" : "false") << '\n';  // true
    vector<int> f2 = {1, 0, 0, 0, 1};
    cout << (canPlaceFlowers(f2, 2) ? "true" : "false") << '\n';  // false
    vector<int> f3 = {0, 0, 1, 0, 0};
    cout << (canPlaceFlowers(f3, 2) ? "true" : "false") << '\n';  // true
    return 0;
}
